### DISCUSSION TITLE
Need help on tacotron2 assertion error on wav file

### DISCUSSION THREAD

**Discussion by JRMeyer** [coqui-ai/TTS#103] [Archived]
<USER>>>> Margarita<br/> <DATE> [October 4, 2019, 11:52pm]<br/> <br/> Hi there!<br/> <br/> I'm using **dev branch** to train **Tacotron2** on a new Chinese tts<br/> dataset. While training, an assertion error occurred and related to one<br/> specific .wav file. *slash* <br/> <br/> 19-31-45.png.filename}[1286×610 133<br/> <br/> I looked into that wav file, it turns out to include only numbers, and<br/> the corresponding text looks like '5123489...', which might be the case<br/> but I don't know how to fix it.<br/> <br/> However, I trained **tacotron** of **master branch** several days<br/> before, and it was all good.<br/> <br/> **config.json (master tacotron):**<br/> <br/> { *slash* <br/> 'run_name': 'zc360_wavenet', *slash* <br/> 'run_description': 'gradual training with prenet frame size 1 + no<br/> maxout for cbhg + symmetric norm.',<br/> <br/> 'audio':{<br/> // Audio processing parameters<br/> 'num_mels': 80, // size of the mel spec frame.<br/> 'num_freq': 1025, // number of stft frequency levels. Size of the linear spectogram frame.<br/> 'sample_rate': 16000, // DATASET-RELATED: wav sample-rate. If different than the original data, it is resampled.<br/> 'frame_length_ms': 50, // stft window length in ms.<br/> 'frame_shift_ms': 12.5, // stft window hop-lengh in ms.<br/> 'preemphasis': 0.98, // pre-emphasis to reduce spec noise and make it more structured. If 0.0, no -pre-emphasis.<br/> 'min_level_db': -100, // normalization range<br/> 'ref_level_db': 20, // reference level db, theoretically 20db is the sound of air.<br/> 'power': 1.5, // value to sharpen wav signals after GL algorithm.<br/> 'griffin_lim_iters': 60,// #griffin-lim iterations. 30-60 is a good range. Larger the value, slower the generation.<br/> // Normalization parameters<br/> 'signal_norm': true, // normalize the spec values in range [0, 1]<br/> 'symmetric_norm': true, // move normalization to range [-1, 1]<br/> 'max_norm': 4, // scale normalization to range [-max_norm, max_norm] or [0, max_norm]<br/> 'clip_norm': true, // clip normalized values into the range.<br/> 'mel_fmin': 0.0, // minimum freq level for mel-spec. ~50 for male and ~95 for female voices. Tune for dataset!!<br/> 'mel_fmax': 8000.0, // maximum freq level for mel-spec. Tune for dataset!!<br/> 'do_trim_silence': true // enable trimming of slience of audio as you load it. LJspeech (false), TWEB (false), Nancy (true)<br/> },<br/> <br/> 'distributed':{<br/> 'backend': 'nccl',<br/> 'url': 'tcp: *slash* / *slash* /localhost:54321'<br/> },<br/> <br/> 'reinit_layers': [],<br/> <br/> 'model': 'Tacotron', // one of the model in models/ <br/> 'grad_clip': 1, // upper limit for gradients for clipping.<br/> 'epochs': 1000, // total number of epochs to train.<br/> 'lr': 0.0001, // Initial learning rate. If Noam decay is active, maximum learning rate.<br/> 'lr_decay': false, // if true, Noam learning rate decaying is applied through training.<br/> 'warmup_steps': 4000, // Noam decay steps to increase the learning rate from 0 to 'lr'<br/> 'memory_size': -1, // ONLY TACOTRON - size of the memory queue used fro storing last decoder predictions for auto-regression. If < 0, memory queue is disabled and decoder only uses the last prediction frame.<br/> 'attention_norm': 'sigmoid', // softmax or sigmoid. Suggested to use softmax for Tacotron2 and sigmoid for Tacotron.<br/> 'prenet_type': 'original', // 'original' or 'bn'.<br/> 'prenet_dropout': true, // enable/disable dropout at prenet.<br/> 'windowing': false, // Enables attention windowing. Used only in eval mode.<br/> 'use_forward_attn': false, // if it uses forward attention. In general, it aligns faster.<br/> 'forward_attn_mask': false,<br/> 'transition_agent': false, // enable/disable transition agent of forward attention.<br/> 'location_attn': true, // enable_disable location sensitive attention. It is enabled for TACOTRON by default.<br/> 'loss_masking': true, // enable / disable loss masking against the sequence padding.<br/> 'enable_eos_bos_chars': false, // enable/disable beginning of sentence and end of sentence chars.<br/> 'stopnet': true, // Train stopnet predicting the end of synthesis.<br/> 'separate_stopnet': true, // Train stopnet seperately if 'stopnet==true'. It prevents stopnet loss to influence the rest of the model. It causes a better model, but it trains SLOWER.<br/> 'tb_model_param_stats': false, // true, plots param stats per layer on tensorboard. Might be memory consuming, but good for debugging.<br/> <br/> 'batch_size': 32, // Batch size for training. Lower values than 32 might cause hard to learn attention. It is overwritten by 'gradual_training'.<br/> 'eval_batch_size':16, <br/> 'r': 7, // Number of decoder frames to predict per iteration. Set the initial values if gradual training is enabled. <br/> 'gradual_training': [[0, 7, 32], [10000, 5, 32], [50000, 3, 32], [130000, 2, 16], [290000, 1, 8]], // ONLY TACOTRON - set gradual training steps [first_step, r, batch_size]. If it is null, gradual training is disabled. <br/> 'wd': 0.000001, // Weight decay weight.<br/> 'checkpoint': true, // If true, it saves checkpoints per 'save_step'<br/> 'save_step': 10000, // Number of training steps expected to save traning stats and checkpoints.<br/> 'print_step': 25, // Number of steps to log traning on console.<br/> 'batch_group_size': 0, //Number of batches to shuffle after bucketing.<br/> <br/> 'run_eval': true,<br/> 'test_delay_epochs': 5, //Until attention is aligned, testing only wastes computation time.<br/> 'test_sentences_file': null, // set a file to load sentences to be used for testing. If it is null then we use default english sentences.<br/> 'data_path': '/home/data1/yutao/tts_data/zc360/', // DATASET-RELATED: can overwritten from command argument<br/> 'meta_file_train': 'metadata_train.csv', // DATASET-RELATED: metafile for training dataloader.<br/> 'meta_file_val': 'metadata_val.csv', // DATASET-RELATED: metafile for evaluation dataloader.<br/> 'dataset': 'ljspeech', // DATASET-RELATED: one of TTS.dataset.preprocessors depending on your target dataset. Use 'tts_cache' for pre-computed dataset by extract_features.py<br/> 'min_seq_len': 6, // DATASET-RELATED: minimum text length to use in training<br/> 'max_seq_len': 150, // DATASET-RELATED: maximum text length<br/> 'output_path': '../zc360_keep/', // DATASET-RELATED: output path for all training outputs.<br/> 'num_loader_workers': 4, // number of training data loader processes. Don't set it too big. 4-8 are good values.<br/> 'num_val_loader_workers': 4, // number of evaluation data loader processes.<br/> 'phoneme_cache_path': 'mozilla_us_phonemes', // phoneme computation is slow, therefore, it caches results in the given folder.<br/> 'use_phonemes': true, // use phonemes instead of raw characters. It is suggested for better pronounciation.<br/> 'phoneme_language': 'zh', // depending on your target language, pick one from https://github.com/bootphon/phonemizer#languages<br/> 'text_cleaner': 'phoneme_cleaners',<br/> 'use_speaker_embedding': false, // use speaker embedding to enable multi-speaker learning.<br/> 'style_wav_for_test': null // path to style wav file to be used in TacotronGST inference.<br/> <br/> }<br/> <br/> **config.json (dev tacotron2):**<br/> <br/> { *slash* <br/> 'run_name': 'zc360', *slash* <br/> 'run_description': 'Tacotron prenet fix test run - dev-memory_fix',<br/> <br/> 'audio':{<br/> // Audio processing parameters<br/> 'num_mels': 80, // size of the mel spec frame.<br/> 'num_freq': 1025, // number of stft frequency levels. Size of the linear spectogram frame.<br/> 'sample_rate': 16000, // DATASET-RELATED: wav sample-rate. If different than the original data, it is resampled.<br/> 'frame_length_ms': 50, // stft window length in ms.<br/> 'frame_shift_ms': 12.5, // stft window hop-lengh in ms.<br/> 'preemphasis': 0.98, // pre-emphasis to reduce spec noise and make it more structured. If 0.0, no -pre-emphasis.<br/> 'min_level_db': -100, // normalization range<br/> 'ref_level_db': 20, // reference level db, theoretically 20db is the sound of air.<br/> 'power': 1.5, // value to sharpen wav signals after GL algorithm.<br/> 'griffin_lim_iters': 60,// #griffin-lim iterations. 30-60 is a good range. Larger the value, slower the generation.<br/> // Normalization parameters<br/> 'signal_norm': true, // normalize the spec values in range [0, 1]<br/> 'symmetric_norm': true, // move normalization to range [-1, 1]<br/> 'max_norm': 4, // scale normalization to range [-max_norm, max_norm] or [0, max_norm]<br/> 'clip_norm': true, // clip normalized values into the range.<br/> 'mel_fmin': 0.0, // minimum freq level for mel-spec. ~50 for male and ~95 for female voices. Tune for dataset!!<br/> 'mel_fmax': 8000.0, // maximum freq level for mel-spec. Tune for dataset!!<br/> 'do_trim_silence': true // enable trimming of slience of audio as you load it. LJspeech (false), TWEB (false), Nancy (true)<br/> },<br/> <br/> 'distributed':{<br/> 'backend': 'nccl',<br/> 'url': 'tcp: *slash* / *slash* /localhost:54321'<br/> },<br/> <br/> 'reinit_layers': [],<br/> <br/> 'model': 'Tacotron2', // one of the model in models/ <br/> 'grad_clip': 1, // upper limit for gradients for clipping.<br/> 'epochs': 1000, // total number of epochs to train.<br/> 'lr': 0.0001, // Initial learning rate. If Noam decay is active, maximum learning rate.<br/> 'lr_decay': false, // if true, Noam learning rate decaying is applied through training.<br/> 'warmup_steps': 4000, // Noam decay steps to increase the learning rate from 0 to 'lr'<br/> 'memory_size': -1, // ONLY TACOTRON - size of the memory queue used fro storing last decoder predictions for auto-regression. If < 0, memory queue is disabled and decoder only uses the last prediction frame.<br/> 'attention_norm': 'softmax', // softmax or sigmoid. Suggested to use softmax for Tacotron2 and sigmoid for Tacotron.<br/> 'prenet_type': 'original', // 'original' or 'bn'.<br/> 'prenet_dropout': true, // enable/disable dropout at prenet.<br/> 'windowing': false, // Enables attention windowing. Used only in eval mode.<br/> 'use_forward_attn': false, // if it uses forward attention. In general, it aligns faster.<br/> 'forward_attn_mask': false,<br/> 'transition_agent': false, // enable/disable transition agent of forward attention.<br/> 'location_attn': true, // enable_disable location sensitive attention. It is enabled for TACOTRON by default.<br/> 'loss_masking': true, // enable / disable loss masking against the sequence padding.<br/> 'enable_eos_bos_chars': false, // enable/disable beginning of sentence and end of sentence chars.<br/> 'stopnet': true, // Train stopnet predicting the end of synthesis.<br/> 'separate_stopnet': true, // Train stopnet seperately if 'stopnet==true'. It prevents stopnet loss to influence the rest of the model. It causes a better model, but it trains SLOWER.<br/> 'tb_model_param_stats': false, // true, plots param stats per layer on tensorboard. Might be memory consuming, but good for debugging.<br/> <br/> 'batch_size': 32, // Batch size for training. Lower values than 32 might cause hard to learn attention. It is overwritten by 'gradual_training'.<br/> 'eval_batch_size':16, <br/> 'r': 7, // Number of decoder frames to predict per iteration. Set the initial values if gradual training is enabled. <br/> <br/> // 'gradual_training': *slash* [ *slash* [0, 7, 32 *slash* ], *slash* [1, 5, 32 *slash* ], *slash* [50000, 3, 32 *slash* ],<br/> *slash* [130000, 2, 16 *slash* ], *slash* [290000, 1, 8 *slash* ] *slash* ], // ONLY TACOTRON - set gradual<br/> training steps *slash* [first_step, r, batch_size *slash* ]. If it is null, gradual<br/> training is disabled. *slash* <br/> 'gradual_training': null, *slash* <br/> 'wd': 0.000001, // Weight decay weight. *slash* <br/> 'checkpoint': true, // If true, it saves checkpoints per 'save_step' *slash* <br/> 'save_step': 10000, // Number of training steps expected to save<br/> traninpg stats and checkpoints. *slash* <br/> 'print_step': 25, // Number of steps to log traning on console. *slash* <br/> 'batch_group_size': 0, //Number of batches to shuffle after bucketing.<br/> <br/> 'run_eval': true,<br/> 'test_delay_epochs': 5, //Until attention is aligned, testing only wastes computation time.<br/> 'test_sentences_file': null, // set a file to load sentences to be used for testing. If it is null then we use default english sentences.<br/> 'min_seq_len': 6, // DATASET-RELATED: minimum text length to use in training<br/> 'max_seq_len': 150, // DATASET-RELATED: maximum text length<br/> 'output_path': '../zc360_keep/', // DATASET-RELATED: output path for all training outputs.<br/> 'num_loader_workers': 4, // number of training data loader processes. Don't set it too big. 4-8 are good values.<br/> 'num_val_loader_workers': 4, // number of evaluation data loader processes.<br/> 'phoneme_cache_path': 'mozilla_us_phonemes', // phoneme computation is slow, therefore, it caches results in the given folder.<br/> 'use_phonemes': false, // use phonemes instead of raw characters. It is suggested for better pronounciation.<br/> 'phoneme_language': 'zh', // depending on your target language, pick one from https://github.com/bootphon/phonemizer#languages<br/> 'text_cleaner': 'phoneme_cleaners',<br/> 'use_speaker_embedding': false, // use speaker embedding to enable multi-speaker learning.<br/> 'style_wav_for_test': null, // path to style wav file to be used in TacotronGST inference.<br/> 'use_gst': false, // TACOTRON ONLY: use global style tokens<br/> <br/> 'datasets': // List of datasets. They all merged and they get different speaker_ids.<br/> [<br/> {<br/> 'name': 'ljspeech',<br/> 'path': '/../../data1/yutao/tts_data/zc360_yanli/',<br/> 'meta_file_train': 'metadata_train.csv',<br/> 'meta_file_val': 'metadata_val.csv'<br/> }<br/> ]<br/> <br/> }<br/> <br/> Thanks!<br/> <br/> [This is an archived TTS discussion thread from discourse.mozilla.org/t/need-help-on-tacotron2-assertion-error-on-wav-file]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> nmstoker<br/> <DATE> [October 24, 2019, 1:45am]<br/> <br/> Maybe I'm missing something, but would it not just be easier to remove<br/> the problem .wav file from your dataset?<br/> <br/> Otherwise you could look through the source code of the dataloader in<br/> the torch module and see what the assert is being triggered by.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> Aymen<br/> <DATE> [September 21, 2020, 11:23am]<br/> <br/> Hi,<br/> <br/> changing this line in datasets/TTSdataset.py *slash* <br/> assert text.size *slash* > 0, self.items *slash* [idx *slash* ] *slash* [1 *slash* ] *slash* <br/> To *slash* <br/> assert text.size *slash* >= 0, self.items *slash* [idx *slash* ] *slash* [1 *slash* ] *slash* <br/> resolved my problem<br/> <br/> [Archived Post]<br/>
